### ISSUE TITLE
Undefined constant OCA\\Cospend\\AppInfo\\Application::ACCESS_PARTICIPANT

### ISSUE THREAD

**Issue by francoisfreitag** [eneiluj/cospend-nc#63]
Thanks for the app, it’s extremely useful to me. :star: 

Since the last version (1.3.10), I noticed the following error when trying to create a bill:
```json
{
    "reqId": "QM3qG3iUUT8d73DtgYIx",
    "level": 3,
    "time": "2021-07-23T11:21:54+00:00",
    "remoteAddr": "REDACTED",
    "user": "friend",
    "app": "index",
    "method": "POST",
    "url": "/apps/cospend/projects/compte/bills",
    "message": "Undefined constant OCA\\Cospend\\AppInfo\\Application::ACCESS_PARTICIPANT",
    "userAgent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:90.0) Gecko/20100101 Firefox/90.0",
    "version": "22.0.0.11",
    "exception": {
        "Exception": "Exception",
        "Message": "Undefined constant OCA\\Cospend\\AppInfo\\Application::ACCESS_PARTICIPANT",
        "Code": 0,
        "Trace": [
            {
                "file": "/usr/share/webapps/nextcloud/lib/private/AppFramework/App.php",
                "line": 156,
                "function": "dispatch",
                "class": "OC\\AppFramework\\Http\\Dispatcher",
                "type": "->"
            },
            {
                "file": "/usr/share/webapps/nextcloud/lib/private/Route/Router.php",
                "line": 301,
                "function": "main",
                "class": "OC\\AppFramework\\App",
                "type": "::"
            },
            {
                "file": "/usr/share/webapps/nextcloud/lib/base.php",
                "line": 1000,
                "function": "match",
                "class": "OC\\Route\\Router",
                "type": "->"
            },
            {
                "file": "/usr/share/webapps/nextcloud/index.php",
                "line": 36,
                "function": "handleRequest",
                "class": "OC",
                "type": "::"
            }
        ],
        "File": "/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
        "Line": 158,
        "Previous": {
            "Exception": "Error",
            "Message": "Undefined constant OCA\\Cospend\\AppInfo\\Application::ACCESS_PARTICIPANT",
            "Code": 0,
            "Trace": [
                {
                    "file": "/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 217,
                    "function": "webAddBill",
                    "class": "OCA\\Cospend\\Controller\\PageController",
                    "type": "->"
                },
                {
                    "file": "/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
                    "line": 126,
                    "function": "executeController",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->"
                },
                {
                    "file": "/usr/share/webapps/nextcloud/lib/private/AppFramework/App.php",
                    "line": 156,
                    "function": "dispatch",
                    "class": "OC\\AppFramework\\Http\\Dispatcher",
                    "type": "->"
                },
                {
                    "file": "/usr/share/webapps/nextcloud/lib/private/Route/Router.php",
                    "line": 301,
                    "function": "main",
                    "class": "OC\\AppFramework\\App",
                    "type": "::"
                },
                {
                    "file": "/usr/share/webapps/nextcloud/lib/base.php",
                    "line": 1000,
                    "function": "match",
                    "class": "OC\\Route\\Router",
                    "type": "->"
                },
                {
                    "file": "/usr/share/webapps/nextcloud/index.php",
                    "line": 36,
                    "function": "handleRequest",
                    "class": "OC",
                    "type": "::"
                }
            ],
            "File": "/var/lib/nextcloud/apps/cospend/lib/Controller/PageController.php",
            "Line": 702
        },
        "CustomMessage": "--"
    },
    "id": "60faa6732e96d"
}
```

**Comment by eneiluj** [comment]
Thanks for the bug report. This is very strange. I can't reproduce this bug.

How did you upgrade the app?
Are you logged in when you're adding a bill and facing this issue? Or are you using a public link (or the guest/password access)? I guess you're logged in because your log mentions line 702 of PageController...

Could you check the content of `.../nextcloud/apps/cospend/lib/AppInfo/Application.php`? You should see this at line 61:
``` php
public const ACCESS_PARTICIPANT = 2;
```
and line 15 of `.../nextcloud/apps/cospend/lib/Controller/PageController.php` should be:
``` php
use OCA\Cospend\AppInfo\Application;
```

The weirdest aspect of your problem is that if the constant was actually missing, you shouldn't be able to get the project list, the bill list etc...

Could it be a cache issue on your server side? Could you try to restart your webserver and whatever system you use for caching (redis, anything redis-like :grin:)?

**Comment by francoisfreitag** [comment]
Thanks a lot for your reply, it did solve my issue.

> How did you upgrade the app?

I used the Nextcloud apps menu (`settings/apps`).

> Are you logged in when you're adding a bill and facing this issue?

Yes, I’m logged in with my user.

I restarted the PHP-FPM process and Redis. After the restart, I can no longer reproduce the failure.

---
It’s due to my PHP config :facepalm:.
I setup aggressive PHP settings to squeeze performance out of the Raspberry PI, with [`opcache.validate_timestamps = 0`](https://www.php.net/manual/en/opcache.configuration.php#ini.opcache.validate-timestamps). Of course, an app upgrade is likely to change PHP files and the opcache should be reset.
